### PR TITLE
[8.0][IMP] Performance issues in mail_tracking addons

### PR DESCRIPTION
**Pull request by antespi**
This PR fix some performance issues:
- [Fixed] In some cases `self.env.cr` is not closed properly.
- Add `index=True` to `time` and `recipient_address` fields of `mail.tracking.email` model
- Many2many relation between partners (and mail contacts) and mail tracking emails deleted
- `email_score` is now a non-stored computed field

Also includes some new features for processing bounced emails easier:
- Adds a menu for listing mass mailing emails and events
- Add a boolean field `email_bounced` to partners and mail contacts. To easy filter them and do manual operations: call the partner, remove the partner, opt-out, ... Also prepare easy inherited methods for programming automatic operations in customer addon.

### CODE REVIEW

File: mail_tracking/models/mail_tracking_email.py
Comment (pedrobaeza) [inline mail_tracking/models/mail_tracking_email.py]: Why not make it non stored?

**Comment by antespi** [comment]
We are running this in one production instance, if works well, then I'll add new tests to improve codecov

File: mail_tracking/controllers/main.py
Comment (yajo) [inline mail_tracking/controllers/main.py]: Not sure if you can use [`with env.cr:` or `with env.cr.savepoint():`](https://github.com/odoo/odoo/blob/9.0/openerp/sql_db.py#L385-L414)

**Comment by yajo** [comment]
Could you please add some docstrings and comments, specially in the performance-critical parts?

File: mail_tracking/controllers/main.py
Comment (antespi) [reply mail_tracking/controllers/main.py]: This pattern is equivalent to `with`, but in this case if more explicit and easier to debug if we need it in future. See: http://effbot.org/zone/python-with-statement.htm

Comment (lasley) [reply mail_tracking/controllers/main.py]: I don't see a functional difference though? This pattern [fails Lint](https://travis-ci.org/OCA/social/jobs/167296604#L333) & I am not sure that debugging is a good enough reason to keep it.

File: mail_tracking/models/mail_tracking_email.py
Comment (lasley) [inline mail_tracking/models/mail_tracking_email.py]: s/growth/grow

File: mail_tracking/models/mail_tracking_email.py
Comment (lasley) [inline mail_tracking/models/mail_tracking_email.py:96]: `@api.model`

File: mail_tracking/models/mail_tracking_email.py
Comment (lasley) [inline mail_tracking/models/mail_tracking_email.py]: `elif` - save some resources - if it already matched the 100, it definitely won't be less than 0

File: mail_tracking/models/mail_tracking_email.py
Comment (lasley) [inline mail_tracking/models/mail_tracking_email.py]: This would be cleaner & more efficient in a dictionary:

```
states = {
    'error': -50.0,
    'opened': 5.0,
   ...
}
score += states.get(tracking.state, 0.0)
```

File: mail_tracking/models/mail_tracking_email.py
Comment (lasley) [inline mail_tracking/models/mail_tracking_email.py]: Why can we not use savepoint here?

File: mail_tracking/models/res_partner.py
Comment (lasley) [inline mail_tracking/models/res_partner.py]: AFAIK computed, non-stored attrs without an inverse are implicitly read only

File: mail_tracking_mass_mailing/hooks.py
Comment (lasley) [inline mail_tracking_mass_mailing/hooks.py]: Please remove this pragma ~~& vote on OCA/maintainer-quality-tools#372~~

**Review by lasley — CHANGES_REQUESTED** [review]

File: mail_tracking/models/mail_tracking_email.py
Comment (antespi) [reply mail_tracking/models/mail_tracking_email.py]: You're right

File: mail_tracking/controllers/main.py
Comment (antespi) [reply mail_tracking/controllers/main.py]: Thanks @lasley, I didn't read transaction related guidelines before. In a multi-db instance we need  to get a new env (and cursor) because ORM can initialize it (several databases, none selected). But I can make it better following guidelines. I'm doing it right now

File: mail_tracking_mass_mailing/hooks.py
Comment (antespi) [reply mail_tracking_mass_mailing/hooks.py]: Already merged, thanks :smile:

File: mail_tracking/models/mail_tracking_email.py
Comment (antespi) [reply mail_tracking/models/mail_tracking_email.py]: when using savepoint, if everything is going well (no exceptions), then transaction is committed?
And, if we are in testing mode, transaction is committed too?

File: mail_tracking/models/res_partner.py
Comment (antespi) [reply mail_tracking/models/res_partner.py]: ok

File: mail_tracking/models/mail_tracking_email.py
Comment (lasley) [reply mail_tracking/models/mail_tracking_email.py]: Hmmm ok looking at the intent here, we may not want a Savepoint. It's not immediately committing the transaction, and will roll back entirely if another transaction afterwards fails (assuming that transaction is also not within a savepoint). 

Think of it like a save-guard - if the transaction within the Savepoint fails, it will rollback just within that savepoint. If the transaction fails outside of a savepoint, all mutations & savepoints in that transaction will fail.

Note this is a Postgres feature, so it's significantly easier to play with in direct SQL if you're curious. Surprisingly, the [Postgres docs on Savepoint](https://www.postgresql.org/docs/8.1/static/sql-savepoint.html) are pretty nice & [the Odoo implementation](https://github.com/odoo/odoo/blob/fc2e80cb4bcc450762c7ac5cb82a3e2d88062b38/odoo/sql_db.py#L396) matches well.

Where is the exclusive lock that is mentioned in the comments actually created?

Comment (antespi) [reply mail_tracking/models/mail_tracking_email.py]: I can see this Odoo code: https://github.com/odoo/odoo/blob/d81258ab8e0dcb6c0df3a47ab8084ec99b7d3e24/openerp/sql_db.py#L390
And this PostgreSQL doc: https://www.postgresql.org/docs/9.1/static/sql-release-savepoint.html

According to that, RELEASE SAVEPOINT just remove the savepoint. But in this case, I need to have the event (and mail state) visible to other isolated cursors and release exclusivelock to mail_tracking_email and mail_tracking_event tables. With this commit, then other writes (for instance to res_partner table) can be done after reading mail_tracking_email or mail_tracking_event tables.

Comment (antespi) [reply mail_tracking/models/mail_tracking_email.py]: After thinking twice, I remove this commit. Now the addon is not reading mail_tracking_email nor email_tracking_event tables after create/write on them, so I don't need this.

**Comment by antespi** [comment]
@lasley Thanks for your review. Changes done

File: mail_tracking/controllers/main.py
Comment (lasley) [inline mail_tracking/controllers/main.py:32]: Seeing as we've confirmed that we're awesome and doing things right, we'll need to disable lint on this line (I'm assuming, wait for Travis to find out for sure):

```
new_env.cr.commit()  #pylint: disable=E8102
```

**Review by lasley — APPROVED** [review]
One note on pylint, but we'll wait for Travis for confirmation

**Comment by antespi** [comment]
pylint is cool :smile:

File: mail_tracking/controllers/main.py
Comment (antespi) [reply mail_tracking/controllers/main.py:32]: Confirmed

Comment (pedrobaeza) [reply mail_tracking/controllers/main.py:32]: This check is for avoiding bad commit() calls without control. If you are sure about the use, then we can put the pylint disable sentence.

**Comment by antespi** [comment]
These improvements have been running well in a production instance today, with one mass mailing sending and Mailgun webhooks been received and 6 workers.

**Review by rafaelbn — APPROVED** [review]
OK, thanks

File: mail_tracking/controllers/main.py
Comment (pedrobaeza) [inline mail_tracking/controllers/main.py:53]: Use `search_count` for optimizing this (see the tip in https://www.odoo.com/documentation/8.0/reference/orm.html#common-orm-methods)

File: mail_tracking/models/res_partner.py
Comment (pedrobaeza) [inline mail_tracking/models/res_partner.py]: Don't put string, as it's automatically inferred from the variable name.

File: mail_tracking/models/res_partner.py
Comment (pedrobaeza) [inline mail_tracking/models/res_partner.py]: The same about the string attribute

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: mail_tracking/controllers/main.py
Comment (antespi) [reply mail_tracking/controllers/main.py:53]: I use tracking_email recordset after that

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**
